File: app/src/main/java/com/amte/shcha/fileencryptiontest/MainActivity.java
package com.amte.shcha.fileencryptiontest;

import android.Manifest;
import android.content.ContentResolver;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;

import android.os.Environment;
import android.util.Log;
import android.view.View;

import android.view.Menu;
import android.view.MenuItem;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.security.InvalidKeyException;
import java.security.Key;
import java.security.NoSuchAlgorithmException;
import java.security.Permission;
import java.text.SimpleDateFormat;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    private static final int REQUEST_CODE_PERMISSION     = 100;
    private static final int REQUEST_CODE_ATTACH_VIDEO     = 102;
    private static final String AES_KEY = "dpfxldldkfdpvmdk";
    public static String [] sPermissions = new String [] {
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };
    private static final String FILE_DIR = Environment.getExternalStoragePublicDirectory(
            Environment.DIRECTORY_DCIM).getAbsolutePath();
    private static final int BUFFER_SIZE = 8 * 1024;
    private File saveFile;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent innerIntent = new Intent(Intent.ACTION_GET_CONTENT);
                innerIntent.setType("video/*");
                innerIntent.putExtra(Intent.EXTRA_LOCAL_ONLY, true);
                Intent wrapperIntent = Intent.createChooser(innerIntent, null);
                startActivityForResult(wrapperIntent, REQUEST_CODE_ATTACH_VIDEO);
            }
        });

        if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, sPermissions, REQUEST_CODE_PERMISSION);
        }
//        saveFile = createFile();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        Log.i (TAG, "onRequestPermissionsResult, grantResults = " + grantResults[0]);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode != RESULT_OK) {
            Log.d(TAG, "bail due to resultCode=" + resultCode);
            return;
        }
        Log.i (TAG, "file uri = " + data.getData());

        ContentResolver cr = getContentResolver();
        InputStream input = null;
        FileOutputStream fout = null;

        try {
            input = cr.openInputStream(data.getData());

            Key secureKey = new SecretKeySpec(AES_KEY.getBytes("UTF-8"), "AES");
            Cipher cipher = Cipher.getInstance("AES");
            cipher.init(Cipher.ENCRYPT_MODE, secureKey);

            if (input instanceof FileInputStream) {
                File saveFile = new File(FILE_DIR,"encrypted.mpg");
                Log.d(TAG, "file = " + saveFile.getPath());
                File parentFile = saveFile.getParentFile();
                if (!parentFile.exists() && !parentFile.mkdirs()) {
                    Log.e(TAG, "file out: mkdirs for " + parentFile.getPath() + " failed!");
                }

                FileInputStream fin = (FileInputStream) input;
                fout = new FileOutputStream(saveFile);
                byte[] buffer = new byte[BUFFER_SIZE];
                int size = 0;
                while ( (size = fin.read(buffer)) != -1 ) {
                    cipher.doFinal(buffer);
                    fout.write(cipher.doFinal(buffer), 0, size);
                }

//                Key secureKey = new SecretKeySpec(AES_KEY.getBytes("UTF-8"), "AES");
//                Cipher cipher = Cipher.getInstance("AES");
//                cipher.init(Cipher.ENCRYPT_MODE, secureKey);
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } catch (NoSuchPaddingException e) {
            e.printStackTrace();
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        } catch (InvalidKeyException e) {
            e.printStackTrace();
        } catch (BadPaddingException e) {
            e.printStackTrace();
        } catch (IllegalBlockSizeException e) {
            e.printStackTrace();
        } finally {
            if (input != null) {
                try {
                    input.close();
                } catch (IOException e) {
                    Log.e(TAG, "Exception caught while closing input: ", e);
                }
            }

            if (fout != null) {
                try {
                    fout.close();
                } catch (IOException e) {
                    Log.e(TAG, "Exception caught while closing output: ", e);
                }
            }
        }

    }

    private File createFile() {
        File file = new File(FILE_DIR);
        if(!file.exists()) file.mkdir();

        return new File(FILE_DIR, "encrypted.mpg");
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/amte/shcha/fileencryptiontest/MainActivity.kt
package com.amte.shcha.fileencryptiontest

import android.Manifest
import android.content.ContentValues
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.io.*
import java.security.InvalidKeyException
import java.security.Key
import java.security.NoSuchAlgorithmException
import javax.crypto.BadPaddingException
import javax.crypto.Cipher
import javax.crypto.IllegalBlockSizeException
import javax.crypto.NoSuchPaddingException
import javax.crypto.spec.SecretKeySpec

class MainActivity : AppCompatActivity() {
    private val saveFile: File? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val fab = findViewById<FloatingActionButton>(R.id.fab)
        fab.setOnClickListener {
            val innerIntent = Intent(Intent.ACTION_GET_CONTENT)
            innerIntent.type = "video/*"
            innerIntent.putExtra(Intent.EXTRA_LOCAL_ONLY, true)
            val wrapperIntent = Intent.createChooser(innerIntent, null)
            startActivityForResult(wrapperIntent, REQUEST_CODE_ATTACH_VIDEO)
        }
        if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, sPermissions, REQUEST_CODE_PERMISSION)
        }
        //        saveFile = createFile();
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        Log.i(TAG, "onRequestPermissionsResult, grantResults = " + grantResults[0])
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != RESULT_OK) {
            Log.d(TAG, "bail due to resultCode=$resultCode")
            return
        }
        Log.i(TAG, "file uri = " + data!!.data)
        var savedFileUri = encryptFile(data.data!!)
        Log.i( TAG, "savedFileUri path = ${savedFileUri.path}")
    }

    private fun encryptFile(uri: Uri): Uri {
        var saveFile = File(FILE_DIR, "encrypted2.mpg")
        var input: InputStream? = null
        var fout: FileOutputStream? = null
        try {
            input = contentResolver.openInputStream(uri)
            val secureKey: Key = SecretKeySpec(AES_KEY.toByteArray(charset("UTF-8")),
                    "AES")
            val cipher = Cipher.getInstance("AES")
            cipher.init(Cipher.ENCRYPT_MODE, secureKey)

            if (input is FileInputStream) {

                Log.d(TAG, "file = ${saveFile.path}")
                val parentFile = saveFile.parentFile
                if (!parentFile.exists() && !parentFile.mkdirs()) {
                    Log.e(TAG, "file out: mkdirs for " + parentFile.path + " failed!")
                }
                fout = FileOutputStream(saveFile)
                val buffer = ByteArray(BUFFER_SIZE)
                var size = 0
                while (input.read(buffer).also { size = it } != -1) {
                    fout.write(cipher.doFinal(buffer), 0, size)     // encryption
//                    fout.write(buffer, 0, size)   // just copy
                }

//                Key secureKey = new SecretKeySpec(AES_KEY.getBytes("UTF-8"), "AES");
//                Cipher cipher = Cipher.getInstance("AES");
//                cipher.init(Cipher.ENCRYPT_MODE, secureKey);
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: NoSuchPaddingException) {
            e.printStackTrace()
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
        } catch (e: InvalidKeyException) {
            e.printStackTrace()
        } catch (e: BadPaddingException) {
            e.printStackTrace()
        } catch (e: IllegalBlockSizeException) {
            e.printStackTrace()
        } finally {
            try {
                input?.close()
            } catch (e: IOException) {
                Log.e(TAG, "Exception caught while closing input: ", e)
            }
            try {
                fout?.close()
            } catch (e: IOException) {
                Log.e(TAG, "Exception caught while closing output: ", e)
            }
        }
        return Uri.fromFile(saveFile)
    }

    private fun insertMediaDb(uri: Uri) {
        // to do.. not working

        val values = ContentValues()
//        values.apply {
        values.put(MediaStore.Video.Media.MIME_TYPE, "video/mp4")
        values.put(MediaStore.Video.Media.DATE_ADDED, System.currentTimeMillis())
        values.put(MediaStore.Video.Media.DATE_TAKEN, System.currentTimeMillis())
        values.put(MediaStore.Video.Media.RELATIVE_PATH, uri.path)
        values.put(MediaStore.Video.Media.DISPLAY_NAME, "aaaaaaa")
//        }
        val collection = MediaStore.Video.Media.getContentUri(MediaStore.VOLUME_EXTERNAL)
        Log.d(TAG, " collection = $collection")
        val mediaUri = applicationContext.contentResolver.insert(collection, values)
        Log.d(TAG, "mediaUri = $mediaUri")
    }

    private fun createFile(): File {
        val file = File(FILE_DIR)
        if (!file.exists()) file.mkdir()
        return File(FILE_DIR, "encrypted.mpg")
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId
        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)
    }

    companion object {
        private const val TAG = "MainActivity"
        private const val REQUEST_CODE_PERMISSION = 100
        private const val REQUEST_CODE_ATTACH_VIDEO = 102
        private const val AES_KEY = "dpfxldldkfdpvmdk"
        var sPermissions = arrayOf(
                Manifest.permission.WRITE_EXTERNAL_STORAGE
        )
        private val FILE_DIR = Environment.getExternalStoragePublicDirectory(
                Environment.DIRECTORY_DCIM).absolutePath
        private const val BUFFER_SIZE = 8 * 1024
    }
}